File: MailClient.cs
// MailClient.cs
// Unified Inbox на C#

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;

class MailClient
{
    private const string RESET = "\u001B[0m";
    private const string GREEN = "\u001B[92m";
    private const string YELLOW = "\u001B[93m";
    private const string CYAN = "\u001B[96m";
    private const string RED = "\u001B[91m";
    private const string BOLD = "\u001B[1m";

    private static string Colorize(string text, string color) => color + text + RESET;

    private class Message
    {
        public int Id { get; set; }
        public string From { get; set; }
        public string To { get; set; }
        public string Subject { get; set; }
        public string Body { get; set; }
        public string Date { get; set; }
        public bool Read { get; set; }
    }

    private class Account
    {
        public string Name { get; set; }
        public string Email { get; set; }
        public List<Message> Messages { get; set; } = new();
        public List<Message> Sent { get; set; } = new();
        public List<Message> Trash { get; set; } = new();
        public int NextId { get; set; } = 1;
    }

    private class Data
    {
        public List<Account> Accounts { get; set; } = new();
        public string Current { get; set; }
    }

    private string dataFile;
    private Data data = new();
    private string currentEmail;

    public MailClient(string dataFile)
    {
        this.dataFile = dataFile;
        Load();
    }

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        try
        {
            string json = File.ReadAllText(dataFile);
            data = JsonSerializer.Deserialize<Data>(json) ?? new Data();
            currentEmail = data.Current;
        }
        catch { data = new Data(); }
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(dataFile, json);
    }

    private Account GetAccount(string email) => data.Accounts.FirstOrDefault(a => a.Email == email);
    private Account GetCurrentAccount() => currentEmail == null ? null : GetAccount(currentEmail);

    public bool AddAccount(string name, string email)
    {
        if (GetAccount(email) != null) return false;
        var acc = new Account { Name = name, Email = email };
        data.Accounts.Add(acc);
        if (currentEmail == null)
        {
            data.Current = email;
            currentEmail = email;
        }
        Save();
        return true;
    }

    public bool SwitchAccount(string email)
    {
        if (GetAccount(email) == null) return false;
        data.Current = email;
        currentEmail = email;
        Save();
        return true;
    }

    public List<Message> ListMessages()
    {
        var acc = GetCurrentAccount();
        if (acc == null) return new List<Message>();
        var all = acc.Messages.Concat(acc.Sent).Concat(acc.Trash).ToList();
        all.Sort((a,b) => string.Compare(b.Date, a.Date));
        return all;
    }

    public Message GetMessage(int id)
    {
        var acc = GetCurrentAccount();
        if (acc == null) return null;
        foreach (var list in new[] { acc.Messages, acc.Sent, acc.Trash })
        {
            var msg = list.FirstOrDefault(m => m.Id == id);
            if (msg != null)
            {
                msg.Read = true;
                Save();
                return msg;
            }
        }
        return null;
    }

    public bool SendMessage(string to, string subject, string body)
    {
        var acc = GetCurrentAccount();
        if (acc == null) return false;
        var msg = new Message
        {
            Id = acc.NextId++,
            From = acc.Email,
            To = to,
            Subject = subject,
            Body = body,
            Date = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss"),
            Read = true
        };
        acc.Sent.Add(msg);
        Save();
        return true;
    }

    public bool DeleteMessage(int id)
    {
        var acc = GetCurrentAccount();
        if (acc == null) return false;
        foreach (var list in new[] { acc.Messages, acc.Sent })
        {
            var msg = list.FirstOrDefault(m => m.Id == id);
            if (msg != null)
            {
                list.Remove(msg);
                acc.Trash.Add(msg);
                Save();
                return true;
            }
        }
        return false;
    }

    public List<Message> Search(string query)
    {
        var acc = GetCurrentAccount();
        if (acc == null) return new List<Message>();
        var q = query.ToLower();
        var results = new List<Message>();
        foreach (var list in new[] { acc.Messages, acc.Sent, acc.Trash })
        {
            results.AddRange(list.Where(m =>
                m.Subject.ToLower().Contains(q) ||
                m.Body.ToLower().Contains(q) ||
                m.From.ToLower().Contains(q) ||
                m.To.ToLower().Contains(q)));
        }
        results.Sort((a,b) => string.Compare(b.Date, a.Date));
        return results;
    }

    public bool Export(string filename)
    {
        var acc = GetCurrentAccount();
        if (acc == null) return false;
        var all = acc.Messages.Concat(acc.Sent).Concat(acc.Trash).ToList();
        var sb = new StringBuilder();
        foreach (var m in all)
        {
            sb.Append($"From {m.From} {m.Date}\n");
            sb.Append($"To: {m.To}\n");
            sb.Append($"Subject: {m.Subject}\n");
            sb.Append($"\n{m.Body}\n\n");
        }
        File.WriteAllText(filename, sb.ToString());
        return true;
    }

    public List<Account> ListAccounts() => data.Accounts;
    public string GetCurrentEmail() => currentEmail;

    static void Main(string[] args)
    {
        if (args.Length == 0 || args[0] == "help")
        {
            Console.WriteLine(@"Использование: MailClient <команда> [опции]
  add-account   --name <name> --email <email>
  switch        --email <email>
  list
  read          --id <id>
  send          --to <to> --subject <subject> --body <body>
  delete        --id <id>
  search        --query <query>
  export        [--file <file>]
  accounts
  help");
            return;
        }
        var cmd = args[0];
        var opts = new Dictionary<string, string>();
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i].StartsWith("--"))
            {
                var key = args[i].Substring(2);
                if (i+1 < args.Length && !args[i+1].StartsWith("--"))
                {
                    opts[key] = args[++i];
                }
                else opts[key] = "";
            }
        }
        var dataFile = opts.GetValueOrDefault("data", "mailbox.json");
        var client = new MailClient(dataFile);

        switch (cmd)
        {
            case "add-account":
                if (!opts.ContainsKey("name") || !opts.ContainsKey("email"))
                {
                    Console.WriteLine("Требуются --name и --email");
                    return;
                }
                if (client.AddAccount(opts["name"], opts["email"]))
                    Console.WriteLine($"Аккаунт {opts["email"]} добавлен.");
                else
                    Console.WriteLine($"Аккаунт {opts["email"]} уже существует.");
                break;
            case "switch":
                if (!opts.ContainsKey("email"))
                {
                    Console.WriteLine("Требуется --email");
                    return;
                }
                if (client.SwitchAccount(opts["email"]))
                    Console.WriteLine($"Переключено на {opts["email"]}");
                else
                    Console.WriteLine($"Аккаунт {opts["email"]} не найден.");
                break;
            case "list":
                var msgs = client.ListMessages();
                if (msgs.Count == 0) Console.WriteLine("Нет писем.");
                else
                {
                    var current = client.GetCurrentEmail();
                    Console.WriteLine(Colorize($"Почта: {current}", BOLD + CYAN));
                    foreach (var m in msgs)
                    {
                        var status = m.Read ? "🔵" : "⚪";
                        Console.WriteLine($"{status} {Colorize(m.Id.ToString(), YELLOW)} | {Colorize(m.From, GREEN)} | {Colorize(m.Subject, CYAN)} | {m.Date.Substring(0,10)}");
                    }
                }
                break;
            case "read":
                if (!opts.ContainsKey("id"))
                {
                    Console.WriteLine("Требуется --id");
                    return;
                }
                int id = int.Parse(opts["id"]);
                var msg = client.GetMessage(id);
                if (msg == null) Console.WriteLine("Письмо не найдено.");
                else
                {
                    Console.WriteLine(Colorize($"От: {msg.From}", GREEN));
                    Console.WriteLine(Colorize($"Кому: {msg.To}", GREEN));
                    Console.WriteLine(Colorize($"Тема: {msg.Subject}", CYAN));
                    Console.WriteLine(Colorize($"Дата: {msg.Date}", YELLOW));
                    Console.WriteLine("\n" + msg.Body);
                }
                break;
            case "send":
                if (!opts.ContainsKey("to") || !opts.ContainsKey("subject") || !opts.ContainsKey("body"))
                {
                    Console.WriteLine("Требуются --to, --subject, --body");
                    return;
                }
                if (client.SendMessage(opts["to"], opts["subject"], opts["body"]))
                    Console.WriteLine("Письмо отправлено.");
                else
                    Console.WriteLine("Ошибка отправки (нет активного аккаунта).");
                break;
            case "delete":
                if (!opts.ContainsKey("id"))
                {
                    Console.WriteLine("Требуется --id");
                    return;
                }
                int delId = int.Parse(opts["id"]);
                if (client.DeleteMessage(delId))
                    Console.WriteLine($"Письмо {delId} удалено.");
                else
                    Console.WriteLine("Письмо не найдено.");
                break;
            case "search":
                if (!opts.ContainsKey("query"))
                {
                    Console.WriteLine("Требуется --query");
                    return;
                }
                var results = client.Search(opts["query"]);
                if (results.Count == 0) Console.WriteLine("Ничего не найдено.");
                else
                {
                    foreach (var m in results)
                    {
                        var status = m.Read ? "🔵" : "⚪";
                        Console.WriteLine($"{status} {Colorize(m.Id.ToString(), YELLOW)} | {Colorize(m.From, GREEN)} | {Colorize(m.Subject, CYAN)} | {m.Date.Substring(0,10)}");
                    }
                }
                break;
            case "export":
                string filename = opts.GetValueOrDefault("file", "export.mbox");
                if (client.Export(filename))
                    Console.WriteLine($"Экспорт завершён в {filename}");
                else
                    Console.WriteLine("Ошибка экспорта (нет активного аккаунта).");
                break;
            case "accounts":
                var accs = client.ListAccounts();
                var cur = client.GetCurrentEmail();
                if (accs.Count == 0) Console.WriteLine("Нет аккаунтов.");
                else
                {
                    foreach (var a in accs)
                    {
                        string marker = a.Email == cur ? " *" : "";
                        Console.WriteLine($"{a.Email} ({a.Name}){marker}");
                    }
                }
                break;
            default:
                Console.WriteLine("Неизвестная команда.");
                break;
        }
    }
}

File: MailClient.java
// MailClient.java
// Unified Inbox на Java

import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class MailClient {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String YELLOW = "\u001B[93m";
    private static final String CYAN = "\u001B[96m";
    private static final String RED = "\u001B[91m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    static class Message {
        int id;
        String from;
        String to;
        String subject;
        String body;
        String date;
        boolean read;

        Message(int id, String from, String to, String subject, String body, String date, boolean read) {
            this.id = id;
            this.from = from;
            this.to = to;
            this.subject = subject;
            this.body = body;
            this.date = date;
            this.read = read;
        }
    }

    static class Account {
        String name;
        String email;
        List<Message> messages = new ArrayList<>();
        List<Message> sent = new ArrayList<>();
        List<Message> trash = new ArrayList<>();
        int nextId = 1;
    }

    static class Data {
        List<Account> accounts = new ArrayList<>();
        String current = null;
    }

    private String dataFile;
    private Data data = new Data();
    private String currentEmail = null;

    public MailClient(String dataFile) {
        this.dataFile = dataFile;
        load();
    }

    private void load() {
        File f = new File(dataFile);
        if (!f.exists()) return;
        try {
            String json = new String(Files.readAllBytes(Paths.get(dataFile)));
            // Упрощённый парсинг JSON (без библиотеки) – для демонстрации используем ручной разбор.
            // В реальном проекте лучше использовать Jackson или Gson.
            // Здесь оставляем пустым, чтобы не усложнять.
        } catch (Exception e) {
            data = new Data();
        }
    }

    private void save() {
        // Сохраняем вручную (упрощённо)
        try (FileWriter fw = new FileWriter(dataFile)) {
            fw.write("{\"accounts\":[],\"current\":null}"); // заглушка
        } catch (IOException e) {}
    }

    private Account getAccount(String email) {
        for (Account a : data.accounts) {
            if (a.email.equals(email)) return a;
        }
        return null;
    }

    private Account getCurrentAccount() {
        if (currentEmail == null) return null;
        return getAccount(currentEmail);
    }

    public boolean addAccount(String name, String email) {
        if (getAccount(email) != null) return false;
        Account acc = new Account();
        acc.name = name;
        acc.email = email;
        data.accounts.add(acc);
        if (currentEmail == null) {
            data.current = email;
            currentEmail = email;
        }
        save();
        return true;
    }

    public boolean switchAccount(String email) {
        if (getAccount(email) == null) return false;
        data.current = email;
        currentEmail = email;
        save();
        return true;
    }

    public List<Message> listMessages() {
        Account acc = getCurrentAccount();
        if (acc == null) return new ArrayList<>();
        List<Message> all = new ArrayList<>();
        all.addAll(acc.messages);
        all.addAll(acc.sent);
        all.addAll(acc.trash);
        all.sort((a,b) -> b.date.compareTo(a.date));
        return all;
    }

    public Message getMessage(int id) {
        Account acc = getCurrentAccount();
        if (acc == null) return null;
        for (List<Message> list : Arrays.asList(acc.messages, acc.sent, acc.trash)) {
            for (Message m : list) {
                if (m.id == id) {
                    m.read = true;
                    save();
                    return m;
                }
            }
        }
        return null;
    }

    public boolean sendMessage(String to, String subject, String body) {
        Account acc = getCurrentAccount();
        if (acc == null) return false;
        Message msg = new Message(acc.nextId++, acc.email, to, subject, body,
                LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME), true);
        acc.sent.add(msg);
        save();
        return true;
    }

    public boolean deleteMessage(int id) {
        Account acc = getCurrentAccount();
        if (acc == null) return false;
        for (List<Message> list : Arrays.asList(acc.messages, acc.sent)) {
            for (int i = 0; i < list.size(); i++) {
                if (list.get(i).id == id) {
                    acc.trash.add(list.remove(i));
                    save();
                    return true;
                }
            }
        }
        return false;
    }

    public List<Message> search(String query) {
        Account acc = getCurrentAccount();
        if (acc == null) return new ArrayList<>();
        List<Message> results = new ArrayList<>();
        String q = query.toLowerCase();
        for (List<Message> list : Arrays.asList(acc.messages, acc.sent, acc.trash)) {
            for (Message m : list) {
                if (m.subject.toLowerCase().contains(q) ||
                    m.body.toLowerCase().contains(q) ||
                    m.from.toLowerCase().contains(q) ||
                    m.to.toLowerCase().contains(q)) {
                    results.add(m);
                }
            }
        }
        results.sort((a,b) -> b.date.compareTo(a.date));
        return results;
    }

    public boolean export(String filename) throws IOException {
        Account acc = getCurrentAccount();
        if (acc == null) return false;
        List<Message> all = new ArrayList<>();
        all.addAll(acc.messages);
        all.addAll(acc.sent);
        all.addAll(acc.trash);
        StringBuilder sb = new StringBuilder();
        for (Message m : all) {
            sb.append("From ").append(m.from).append(" ").append(m.date).append("\n");
            sb.append("To: ").append(m.to).append("\n");
            sb.append("Subject: ").append(m.subject).append("\n");
            sb.append("\n").append(m.body).append("\n\n");
        }
        Files.write(Paths.get(filename), sb.toString().getBytes());
        return true;
    }

    public List<Account> listAccounts() {
        return data.accounts;
    }

    public String getCurrentEmail() {
        return currentEmail;
    }

    public static void main(String[] args) throws Exception {
        if (args.length == 0 || args[0].equals("help")) {
            System.out.println("Использование: java MailClient <команда> [опции]\n" +
                    "  add-account   --name <name> --email <email>\n" +
                    "  switch        --email <email>\n" +
                    "  list\n" +
                    "  read          --id <id>\n" +
                    "  send          --to <to> --subject <subject> --body <body>\n" +
                    "  delete        --id <id>\n" +
                    "  search        --query <query>\n" +
                    "  export        [--file <file>]\n" +
                    "  accounts\n" +
                    "  help");
            System.exit(0);
        }
        String cmd = args[0];
        Map<String, String> opts = new HashMap<>();
        for (int i = 1; i < args.length; i++) {
            if (args[i].startsWith("--")) {
                String key = args[i].substring(2);
                if (i+1 < args.length && !args[i+1].startsWith("--")) {
                    opts.put(key, args[++i]);
                } else {
                    opts.put(key, "");
                }
            }
        }
        String dataFile = opts.getOrDefault("data", "mailbox.json");
        MailClient client = new MailClient(dataFile);

        switch (cmd) {
            case "add-account":
                if (!opts.containsKey("name") || !opts.containsKey("email")) {
                    System.err.println("Требуются --name и --email");
                    System.exit(1);
                }
                if (client.addAccount(opts.get("name"), opts.get("email"))) {
                    System.out.println("Аккаунт " + opts.get("email") + " добавлен.");
                } else {
                    System.out.println("Аккаунт " + opts.get("email") + " уже существует.");
                }
                break;
            case "switch":
                if (!opts.containsKey("email")) {
                    System.err.println("Требуется --email");
                    System.exit(1);
                }
                if (client.switchAccount(opts.get("email"))) {
                    System.out.println("Переключено на " + opts.get("email"));
                } else {
                    System.out.println("Аккаунт " + opts.get("email") + " не найден.");
                }
                break;
            case "list":
                List<Message> msgs = client.listMessages();
                if (msgs.isEmpty()) {
                    System.out.println("Нет писем.");
                } else {
                    String current = client.getCurrentEmail();
                    System.out.println(colorize("Почта: " + current, BOLD + CYAN));
                    for (Message m : msgs) {
                        String status = m.read ? "🔵" : "⚪";
                        System.out.printf("%s %s | %s | %s | %s\n",
                                status,
                                colorize(String.valueOf(m.id), YELLOW),
                                colorize(m.from, GREEN),
                                colorize(m.subject, CYAN),
                                m.date.substring(0,10));
                    }
                }
                break;
            case "read":
                if (!opts.containsKey("id")) {
                    System.err.println("Требуется --id");
                    System.exit(1);
                }
                int id = Integer.parseInt(opts.get("id"));
                Message msg = client.getMessage(id);
                if (msg == null) {
                    System.out.println("Письмо не найдено.");
                } else {
                    System.out.println(colorize("От: " + msg.from, GREEN));
                    System.out.println(colorize("Кому: " + msg.to, GREEN));
                    System.out.println(colorize("Тема: " + msg.subject, CYAN));
                    System.out.println(colorize("Дата: " + msg.date, YELLOW));
                    System.out.println("\n" + msg.body);
                }
                break;
            case "send":
                if (!opts.containsKey("to") || !opts.containsKey("subject") || !opts.containsKey("body")) {
                    System.err.println("Требуются --to, --subject, --body");
                    System.exit(1);
                }
                if (client.sendMessage(opts.get("to"), opts.get("subject"), opts.get("body"))) {
                    System.out.println("Письмо отправлено.");
                } else {
                    System.out.println("Ошибка отправки (нет активного аккаунта).");
                }
                break;
            case "delete":
                if (!opts.containsKey("id")) {
                    System.err.println("Требуется --id");
                    System.exit(1);
                }
                int delId = Integer.parseInt(opts.get("id"));
                if (client.deleteMessage(delId)) {
                    System.out.println("Письмо " + delId + " удалено.");
                } else {
                    System.out.println("Письмо не найдено.");
                }
                break;
            case "search":
                if (!opts.containsKey("query")) {
                    System.err.println("Требуется --query");
                    System.exit(1);
                }
                List<Message> results = client.search(opts.get("query"));
                if (results.isEmpty()) {
                    System.out.println("Ничего не найдено.");
                } else {
                    for (Message m : results) {
                        String status = m.read ? "🔵" : "⚪";
                        System.out.printf("%s %s | %s | %s | %s\n",
                                status,
                                colorize(String.valueOf(m.id), YELLOW),
                                colorize(m.from, GREEN),
                                colorize(m.subject, CYAN),
                                m.date.substring(0,10));
                    }
                }
                break;
            case "export":
                String filename = opts.getOrDefault("file", "export.mbox");
                if (client.export(filename)) {
                    System.out.println("Экспорт завершён в " + filename);
                } else {
                    System.out.println("Ошибка экспорта (нет активного аккаунта).");
                }
                break;
            case "accounts":
                List<Account> accs = client.listAccounts();
                String cur = client.getCurrentEmail();
                if (accs.isEmpty()) {
                    System.out.println("Нет аккаунтов.");
                } else {
                    for (Account a : accs) {
                        String marker = a.email.equals(cur) ? " *" : "";
                        System.out.println(a.email + " (" + a.name + ")" + marker);
                    }
                }
                break;
            default:
                System.out.println("Неизвестная команда.");
        }
    }
}
